### PR TITLE
UCT/CUDA-IPC: khash for cuda-ipc memhandles

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
- using khash to avoid repeated memhandle open cost
- removing unused p2p_map member

### CODE REVIEW

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

**Comment by Akshay-Venkatesh** [comment]
@yosefe @bureddy @brminich Please review when possible. I've replaced sglib with khash and have closed #2593

**Comment by bureddy** [comment]
ok to test

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7133/ for details (Mellanox internal link).

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.h [hunk 1]
```diff
@@ -9,14 +9,44 @@
 #include <uct/api/uct.h>
 #include <uct/base/uct_iface.h>
 #include <ucs/type/class.h>
+#include <ucs/datastruct/khash.h>
+#include "cuda_ipc_md.h"
 
+typedef struct uct_cuda_ipc_rem_seg uct_cuda_ipc_rem_seg_t;
+
+struct uct_cuda_ipc_rem_seg {
+    uct_cuda_ipc_rem_seg_t *next;
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: not needed

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.h [hunk 2]
```diff
@@ -9,14 +9,44 @@
...
+static inline khint_t uct_cuda_ipc_memh_hash_func(CUipcMemHandle seg)
+{
+    int hash_val = 7;
+    int i;
+
+    for (i = 0; i < sizeof(seg); i++) {
+        hash_val = hash_val*31 + seg.reserved[i];
+    }
+
+    return (khint_t) (hash_val);
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: casting probably notrneeded here

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.h [hunk 3]
```diff
@@ -9,14 +9,44 @@
...
+    int                    dev_num;    /* GPU Device number */
+};
+
+static inline khint_t uct_cuda_ipc_memh_hash_func(CUipcMemHandle seg)
+{
+    int hash_val = 7;
+    int i;
+
+    for (i = 0; i < sizeof(seg); i++) {
+        hash_val = hash_val*31 + seg.reserved[i];
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: spaces around '*'

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.h [hunk 4]
```diff
@@ -9,14 +9,44 @@
...
+    }
+
+    return (khint_t) (hash_val);
+}
+
+#define uct_cuda_ipc_memh_hash_equal(sg1, sg2)  \
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: pls use _ for macro arguments (eg _sg1 instead of sg1)

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.h [hunk 5]
```diff
@@ -9,14 +9,44 @@
...
+    strncmp((const char *) &sg1, (const char *) &sg2, sizeof(CUipcMemHandle))
```
Comment (yosefe) [inline src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: wrap macro args with ( )

**Comment by yosefe** [comment]
tests failed due to out-of-memory
bot:mlx:retest

File: src/uct/cuda/cuda_ipc/cuda_ipc_ep.h (hunk 2 above)
Comment (Akshay-Venkatesh) [reply src/uct/cuda/cuda_ipc/cuda_ipc_ep.h]: @yosefe I've fixed this. Does the hash function logic look correct to you? Is this needed or is there something simpler that works? From the code, you can see that I use the key just for comparison but unfortunately, the key is a string of a fixed size else if it was an integer, I'd simply use an identity function. Lmk what you think.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7137/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7135/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/7138/ for details (Mellanox internal link).

**Pull request merged**